feat(hc): Publish UserIP logs to kafka from REGION silos

corps opened the pull request:
Testable in development.  You don't need to technically run the CONTROL and REGION servers at the same time, merely run the server as REGION and then `sentry run region-to-silo-consumer --region-name _local`

corps committed:
feat(middleware): FF enabled UserIP -> Kafka

For regional silo deployments, UserIP entries will need to reach control
silo.  To this end, introducing the concept of a per regional silo Kafka
consumer worker and producer that acts as a bus for these to persist in
the control silo.  Behind a sentry_options value which is false, or the
SiloMode which must be set to REGIONAL.

mikejihbe commented on src/sentry/models/userip.py:
```diff
@@ -26,21 +32,53 @@ class Meta:
+        values.update({"country_code": geo["country_code"], "region_code": geo["region"]})
 
+    user.get_orgs()
+    if SiloMode.get_current_mode() == SiloMode.REGION:
+        _perform_kafka_log(user, ip_address, values)
+    else:
         UserIP.objects.create_or_update(user=user, ip_address=ip_address, values=values)
```
@RyanSkonnord was considering an abstraction that would handle this switching for us, which might be a good idea. 

One abstraction I've used before is a function that is both the dispatcher and the handler of an event, so you can call it directly in a context like this `create_or_update_user_ip()` to enqueue the kafka message and the kafka consumer implementation looks identical `create_or_update_user_ip()` but does the actual work because it's in the right silo. It makes for quite readable code, but does hide some complexity 🤷

corps replied:
I don't have a strong opinion, but I think it will be easier to form a better opinion _after_ I add more of these region -> control silo code paths, because I think there will be more complexity in the audit log code paths.

Also, to be clear, is there some general concern having the control/region silos behave differently generally?  I expect both code paths to be tested and neither to be particularly complex.

corps commented:
Ok, finished first pass of consumer, writing tests and making a few more adjustments.

corps commented on src/sentry/api/endpoints/project_release_files.py:
```diff
@@ -120,7 +120,7 @@ def post_releasefile(request, release, logger):
             dist = release.add_dist(dist_name)
 
-        # Quickly check for the presence of this file before continuing with
+        # Quickly check for the region_to_control of this file before continuing with
```
whoops, this appears to be an aggressive IntelliJ rename issue , fixing

corps commented on src/sentry/conf/server.py:
```diff
@@ -2488,6 +2489,8 @@ def build_cdc_postgres_init_db_volume(settings):
     # Metrics Testing Topics
     KAFKA_SNUBA_GENERICS_METRICS_CS: {"cluster": "default"},
+    # User Ip Events -- bucketed by region name
+    KAFKA_USER_IP_EVENTS: {"cluster": {"_local": "default"}},
```
@markstory @mikejihbe 
This is a new concept -- since the consumer of the new topic will be divided by region_name, I needed to be able to embed the extra heirarchy for configuring the cluster name.  I settled on this approach as the path of least resistance, see the code changes below related to getting kafka configuration.

mikejihbe replied:
it looks like @AniketDas-Tekky 's work on https://github.com/getsentry/sentry/pull/38830 doesn't yet include a preliminary region configuration file, so I don't know if we have a format for it yet. I'm guessing we do not. Regardless, I think we might need a dev config file that would define this `_local` name, right?

corps replied:
We shouldn't need any other changes than already in the PR.  "_local" becomes a special region name that maps to the existing behavior (where all clusters are `default`).

corps commented on src/sentry/region_to_control/producer.py:
```diff
@@ -1,28 +1,24 @@
 
-def produce_user_ip(user: User, ip_address: str, values: Dict[str, Any]):
+def produce_user_ip(event: UserIpEvent):
+    values = dataclasses.asdict(event)
```
@markstory @mikejihbe 
So this is where I landed -- in this way, generally the difference between control and regional behavior is non existent because the message matches the schema being posted.  However, and to your point mike, the consumer side needs to be more careful to be able to support 'out of date messages' during rollouts and deploys.  For this reason consumer code path should not be converged with the producer code path in case any repackaging or defaulting is necessary to upgrade old kafka message versions into the new schema.

The problem of general schema evolution can be solved by other means, such as again via the approach Avro takes where 'schema upgrade' metadata becomes _part of the messaging protocol_.  

But, that said, the simplest solution, what I'm doing here, is assuming developers write robust upgrade code.  To facilitate this, I intend to create tests that use fixed payloads in the consumer tests and advice in the comments that developers duplicate and append new tests for new payload schemas, rather than update existing tests.

corps commented on src/sentry/utils/kafka.py:
```diff
@@ -65,7 +66,16 @@ def create_batching_kafka_consumer(topic_names, worker, **options):
+        cluster_names: Set[str] = set()
+        for topic_name in topic_names:
+            cluster_name = settings.KAFKA_TOPICS[topic_name]
+            if region_name is not None:
```
@RyanSkonnord @mikejihbe 

I'm assuming that in the end, the full list of region names and their mapping is a configuration that ops owns, and gets templated into `server.py` overrides.  To support minimal code divergence, I scoped atleast the kafka cluster config by an additional level (region name) which defaults to `"_local"` in the `sentry` repo, and would get templated in production.

I assume we can do a similar approach for other configurations, and that a master list of regions can be an additional `setting` option which also gets templated (and defaults to just `_local` otherwise).

Thoughts?

RyanSkonnord replied:
That sounds reasonable enough for now. There's a good chance we'll have to rewrite a lot of this, however. One of my first steps would be to build some kind of object representation for the region -- the less we have to pass around `region_name` as a magic string, the better.

It seems like it's no longer convenient for the region-mapping config to be theoretical. I'll sketch out a skeleton some time this week, or we could pair on it if you want.

corps replied:
I'm realizing now too that our use cases don't line up precisely. I do actually need to "leave" the magic string argument, because the region specific consumer configs are for the benefit of the control silo which consumes those regions  In other words, the region_name for consumer is not indicative of -this- region, but rather a -origin- region.

corps replied:
This all simplifies away now.

RyanSkonnord commented on src/sentry/region_to_control/producer.py:
```diff
@@ -0,0 +1,33 @@
+user_ip_publisher: Optional[KafkaPublisher] = None
+
+
+def get_user_ip_publisher() -> KafkaPublisher:
```
This could use a comment explaining the motive for caching a global `KafkaPublisher` instance. As it is, it looks like we're just trying to avoid loading the options more than once. Is there more to it, like reusing the same connection or something like that?

If we're maintaining a global instance but initializing it lazily, I might prefer to make it as a `cached_property` of a singleton object. Otherwise, the global variable should be named `_user_ip_publisher` to indicate that people should call `get_user_ip_publisher()` instead of importing `user_ip_publisher`.

markstory commented on src/sentry/conf/server.py:
```diff
@@ -2492,6 +2493,8 @@ def build_cdc_postgres_init_db_volume(settings):
     # Metrics Testing Topics
     KAFKA_SNUBA_GENERICS_METRICS_CS: {"cluster": "default"},
+    # User Ip Events -- bucketed by region name
+    KAFKA_USER_IP_EVENTS: {"cluster": {"_local": "default"}},
```
Couldn't we configure each region with the relevant cluster, or always use the same cluster name and have the host/ip of that cluster name resolve differently to the regional brokers?

mikejihbe replied:
I definitely like the idea of using DNS for these mappings, we just need to make sure XDS will work with this approach (but it sounded like it would.

corps replied:
That's a plausible approach, yes.  It wasn't yet clear as I was writing this how much we could rely on XDS or our service discovery mechanisms, or that the DNS names were consistent enough to ensure a simple mapping between region name -> endpoint.

Furthermore, for me to bake this assumption in does require generally greater complexity in existing kafka consumer logic.  My strategy is to `defer` for now.  I expliitly plan to be looking into the ops side of this equation and figuring out our path forward / maintaining changes in the face of improvements, but I didn't want to overload assumptions here from the git go.

Given that right now we do not but one region + developlment, would it be acceptable to continue down this path of least resistance until I can ascertain the larger picture?

corps replied:
Going forward with this static configuration for now, until the story for how and if we can leverage  XDS or similar to make this more autoconfigurable.

mikejihbe commented on src/sentry/runner/commands/run.py:
```diff
@@ -551,6 +552,38 @@ def ingest_consumer(consumer_types, all_consumer_types, **options):
+    "region_name",
+    "--region-name",
+    required=True,
+    help="Regional name to run the consumer for",
```
We had discussed writing to the local region kafka and pulling from the control silo (for better cross region network resiliency)

corps commented on src/sentry/region_to_control/messages.py:
```diff
@@ -0,0 +1,53 @@
+class UserIpEvent:
+    user_id: int = -1
+    ip_address: str = "127.0.0.1"
+    last_seen: datetime.datetime = datetime.datetime(2000, 1, 1)
```
So, my intuition actually runs counter a bit. 

 I believe having required fields here is more problematic than not.
While it isn't obvious in _this_ PR where the schema and code are tightly coupled for the time being, the _pattern_ that I'm thinking about is when more types of messages are being relayed from region to control, and there are potential schema changes over time with regards to that.  Sticking kafka between the source and consumer of an event has interesting consequences; namely the general requirement that if the write into kafka succeeds, the consumer should be resilient to processing that event at _any point in the future_, including when the schema to produce said values has changed (potentially removing or adding fields).  See below how `Message` objects are constructed from the raw json payloads.

By always having defaults in general, it means that schema upgrades and graceful handling is more natural -- in no case can reconstituting a value result in missing constructor args.  

As for the defaults themselves, the user_id, ip, and last_seen were picked to clearly be distinct from production values.  I'd feel more comfortable with simply changing ip_address default to `""` and adding an explicit check for `user_id` before insert than to remove the defaults.  What do the two of you think of that compromise?

I realize it's a rather theoritical counter argument, but it's baked by experience and intuition of building persisted event consumers.

corps replied:
Talked to @markstory and came up with a compromise for normalization that retains required attributes on producer side.

corps commented on src/sentry/region_to_control/consumer.py:
```diff
@@ -0,0 +1,79 @@
+
+    for topic in topic_names:
+        if topic not in REGION_TO_CONTROL_TOPIC_NAMES:
+            raise ValueError(f"region_to_control_consumer received unexpected topic name: {topic}")
```
Hmm,

I saw that the ingest consumer does the same thing, but as I dug into it I felt that the abstraction wasn't doing much in truth.  

Basically, I'm unclear what introducing the Enum actually helps with.  The enum is functionally quivalent to the set of strings but requires more boilerplate and maintenance (mapping those enums to strings).  It also introduces the potential for drift in the face of renames (both the translated and enum names must both be renamed, but there is no feedback when those two slight).

Using the set of strings goes straight to the heart of the problem with minimal overhead, makes it really obvious what the kafka topic name <-> CLI option mapping is for introspection (always a simple translation), and makes a clear error in the case of rename drift (the option set must map to valid KAFKA_TOPIC entries from which they originate).

corps replied:
This goes away as I reduced the consumer to a single topic for simplicity.

corps commented on src/sentry/region_to_control/consumer.py:
```diff
@@ -0,0 +1,79 @@
+        getattr(settings, topic_name.upper().replace("-", "_")) for topic_name in topic_names
+    }
+
+    return create_batching_kafka_consumer(
```
Roger, I'll take a look at that this afternoon.

corps commented on src/sentry/region_to_control/consumer.py:
```diff
@@ -0,0 +1,79 @@
+    def _flush_batch(self, batch: Sequence[Message]):
+        for row in batch:
+            if row.user_ip_event:
+                updated, created = UserIP.objects.create_or_update(
```
There is already a test covering that exact point.  There is no integrity error thrown -- if you chek the implementation, create_or_update already catches that and simply adjusts the returned `updated` and `created`.

corps commented:
@markstory @RyanSkonnord 

Recently updated, mind taking another pass?

1.  Simplified topics to just one broad region -> control topic which shares all messages.  This is simplest given the 'weird' nature of this kafka topic (consumed by the control silo). 
2.  Moved to using the 'new' kafka consumer approach via arroyo.
3.   Introduced the `SENTRY_REGION_NAME` setting which points to the current region name.  This possibly goes away, again, if we decide to, say, split consumer / producer configurations for kafka.  (As it stands, the consumer of this topic needs regional silo perspectives, but the producer does not.)

corps commented on src/sentry/conf/server.py:
```diff
@@ -2500,6 +2501,8 @@ def build_cdc_postgres_init_db_volume(settings):
     # Metrics Testing Topics
     KAFKA_SNUBA_GENERICS_METRICS_CS: {"cluster": "default"},
+    # User Ip Events -- bucketed by region name
+    KAFKA_REGION_TO_CONTROL: {"cluster": {"_local": "default"}},
```
Actually, I found out that there is a way to split consumer from producer configuration here, so this whole thing can be changed and simplified.  coming in next commit

corps commented on src/sentry/conf/server.py:
```diff
@@ -2816,3 +2819,4 @@ def build_cdc_postgres_init_db_volume(settings):
 SENTRY_PERFORMANCE_ISSUES_RATE_LIMITER_OPTIONS = {}
 
 SENTRY_REGION_CONFIG: Mapping[str, Region] = {}
+SENTRY_REGION_NAME: str = "_local"
```
Gitting rid of this, easier way around it.

corps commented:
Fixing tests in CI, I just needed to create a new job for the kafka consumer test.

corps committed:
Region to control integration test configuration

asottile-sentry commented on .github/workflows/backend.yml:
```diff
@@ -324,6 +324,32 @@ jobs:
         run: |
           make test-plugins
 
+  region-to-control:
```
is this a required check? as written it is not and so a merge will succeed with this failing

can these tests be added to an existing job rather than making a new special one?

corps replied:
Good question.  The main reason this one test needs to be separate from the other backend jobs it he requirement of kafka.  However, it does not cleanly fit into other sections (relay, etc) that are more specialized.  I'd love to consolidate, but it didn't seem to work.

I can add required to the check.

asottile-sentry replied:
hmmm interesting

I wonder if we can't just run kafka for everything -- it's kind of a pain to figure out the right set of things to run tests locally already.  adding one to the number isn't the worst but I wonder if there's a way to land this without doing that

corps replied:
🤷  I mean, I don't disagree.  But conversely, I don't know, other than having a big conversation with dev-infra, how to create consensus to pass this through the eye of the needle.  It's in that grey area where you could argue multiple solutions to the same problem (enable kafka in all test environments, widen an existing test group, adding this test group) such that you might bike shed a lot without a real conclusion.

Considering that changing this after the fact is trivial, and there isn't substantive impact at this point one way or another, I'm really just looking for any definitive answer without too much bike sheding.

markstory replied:
The `snuba` test jobs already spin up kafka. These tests could be added to that build group if you didn't want to add another one./m

corps replied:
Question then: should I also -rename- the test category -and- the make target?  Since it would become a broader category and adding this to that would be a misnomer.  In that case, it also means that existing users may need to "relearn" which make target to use to test snuba.

I guess, honestly, I'm skeptical that making a name or categorical change is attractive compared to adding the new category.  For instance, the category splitting by subject is attractive because changes will tend to break things along specific categories, and rerunning tests within a narrow category is attractive compared to larger ones.

asottile-sentry commented on tests/sentry/region_to_control/test_region_to_control_kafka.py:
```diff
@@ -0,0 +1,81 @@
+    return Factories.create_user("admin@localhost")
+
+
+@pytest.fixture
+def random_group_id():
+    return f"test-consumer-{random.randint(0, 2 ** 16)}"
```
random is usually a poor choice especially in tests -- it leads to inherent flakiness.  can this be a deterministic value instead?

corps replied:
Sure, in this case however the randomness helps reduce false positives in the case of partitioning state influence. This follows the pattern of ingest kafka tests that generate unique group ids.  To be honest, it probably doesn't matter in this singular simple test, but in future tests it could.  My main motivation for doing so was the follow the practice of the bigger kafka test suites now ahead of future growth and changes.

asottile-sentry replied:
if there's false positives then I don't think randomness helps you -- it just makes them flaky false positives rather than deterministic

or perhaps I'm not understanding -- could you elaborate?

corps replied:
Sure, but to be honest, given that I'm introducing just a singular test and I'm riding on a simple implementation for now, I'll accept the feedback and remove the random group id.  The main motivation for including it in the first was following the convention of other kafka consumer integration tests but in this case, I have a singular test and not much else so it isn't likely necessary.  When further work warrants it I can revisit it.

corps committed:
Adding region-to-control to required jobs

corps committed:
Test fixes

corps committed:
Merge branch 'master' into zc/publish-user-ips

corps commented:
Remaining test failure seems to be related to performance flakey test.  Final pass @markstory @RyanSkonnord ?

corps committed:
Merge branch 'master' into zc/publish-user-ips

corps committed:
Merge branch 'master' into zc/publish-user-ips

corps committed:
Merge branch 'master' into zc/publish-user-ips

corps committed:
Merge branch 'master' into zc/publish-user-ips

markstory commented on src/sentry/conf/server.py:
```diff
@@ -2504,6 +2505,8 @@ def build_cdc_postgres_init_db_volume(settings):
     KAFKA_INGEST_REPLAYS_RECORDINGS: {"cluster": "default"},
     # Metrics Testing Topics
     KAFKA_SNUBA_GENERICS_METRICS_CS: {"cluster": "default"},
+    # User Ip Events -- bucketed by region name
```
```suggestion
    # Region to Control Silo messaging - eg. User Ip Events; bucketed by region name
```
Might be more future proof :shrug:

markstory commented on src/sentry/region_to_control/consumer.py:
```diff
@@ -0,0 +1,113 @@
+
+    def create_with_partitions(
+        self,
+        commit: Callable[[Mapping[Partition, Position]], None],
```
```suggestion
        commit: Commit,
```

Should this be using the typevar defined earlier?

corps replied:
Yes

markstory reviewed (commented):
Looking good. Thank you for converting the consumer to arroyo :clap:

corps committed:
Responding to commits

corps committed:
Merge branch 'master' into zc/publish-user-ips

getsantry[bot] committed:
style(lint): Auto commit lint changes

corps committed:
Test fix

corps committed:
Merge branch 'master' into zc/publish-user-ips

corps committed:
Merge branch 'master' into zc/publish-user-ips

markstory reviewed (approved):
Code looks good to me. I don't have strong opinions on how the ci jobs should be configured though.

corps committed:
Test fix

wedamija commented on tests/sentry/ingest/ingest_consumer/test_ingest_consumer_kafka.py:
```diff
@@ -108,7 +107,7 @@ def test_ingest_consumer_reads_from_topic_and_calls_celery_task(
             max_batch_size=2,
             max_batch_time=5000,
-            group_id=random_group_id,
+            group_id="some-group-id",
```
Why'd we change this btw? I'm not sure if it's the reason, but this test has been flakey since yesterday.

wedamija replied:
https://sentry.io/organizations/sentry/issues/3616300100/

corps replied:
Ironically I made the change in response to some review feedback suggesting not to use random group ids to avoid flaky tests, but I did so unfortunately quickly and thoughtlessly.

corps replied:
https://github.com/getsentry/sentry/pull/39429

wedamija replied:
I'm not actually sure whether this is the cause of flakiness, but we'll see if it fixes things afterwards
